### PR TITLE
Added Requester object to handle all HTTP requests

### PR DESCRIPTION
**Pull request by BaptisteGerondeau**

### CODE REVIEW

File: library/common.py [hunk 1]
```diff
@@ -5,25 +5,120 @@
...
 from urllib.parse import quote
 
-class ProvisionerError(Exception):
-    def __init__(self, message):
-        super(ProvisionerError, self).__init__(message)
-
 # TODO: This needs a more standard name (static method)
 # Or we need to make sure the IPGetter object is easy to use
-def get_machine_by_name(mrp_token, mrp_url, machine_name):
+def get_machine_by_name(requester, machine_name):
```
Comment (rengolin) [inline library/common.py]: Nit: "requester" is a horrible name. :)

Something like "URLhandler" would be better.

File: library/common.py [hunk 2]
```diff
@@ -5,25 +5,120 @@
...
-       raise ProvisionerError('Error no assigned machine found with name "{}"'.
-                    format(machine_name))
-    if len(r.json()) > 1:
-       raise ProvisionerError('Error more than one machine found with name "{}", {}'.
-                    format(machine_name, r.json()))
-    return r.json()[0]
+    path = "/api/v1/machine?q={}&show_all=false".format(q)
+    try:
+        result = requester.get(path)
+    except Exception as err:
```
Comment (rengolin) [inline library/common.py]: No need to catch & raise, at least not at this stage, for whatever was the reason before.

If you catch, you have to do something with it.

File: library/common.py [hunk 3]
```diff
@@ -5,25 +5,120 @@
...
+        raise err
+
+    return result[0]
+
+class ProvisionerError(Exception):
+    def __init__(self, message):
+        super(ProvisionerError, self).__init__(message)
+
+class ClientError(Exception):
```
Comment (rengolin) [inline library/common.py]: This is unused at this stage, please remove.

File: library/common.py [hunk 4]
```diff
@@ -5,25 +5,120 @@
...
+class RequesterJSONError(RequesterError):
+    def __init__(self, method, url, response):
+        super(RequesterJSONError, self).__init__("JSON ERROR on %s request at %s" %
+                                                 (str(method), str(url)), method, url)
+        self.response = response
+
+
+class Requester(object):
+    def __init__(self, mrp_url, mrp_token):
+        self.base_url = mrp_url
```
Comment (rengolin) [inline library/common.py]: validation on these fields would make for better error reports.

File: library/get_ip.py [hunk 5]
```diff
@@ -4,63 +4,63 @@
...
 class IPGetter(object):
-    def __init__(self, mrpurl, mrptoken, machine_name, interface_name='eth1'):
-        self.mrp_url = mrpurl
-        self.mrp_token = mrptoken
+    def __init__(self, requester, machine_name, interface_name='eth1'):
+        self.requester = requester
         self.interface = interface_name
         self.machine_id = get_machine_by_name(
-            self.mrp_token, self.mrp_url, machine_name)['id']
+            self.requester, machine_name)['id']
```
Comment (rengolin) [inline library/get_ip.py]: Nit: probably fits all in one line?

File: library/get_ip.py [hunk 6]
```diff
@@ -4,63 +4,63 @@
...
-        if r.status_code != 200:
-            raise ProvisionerError('Error fetching {} {}, HTTP {} {}'.format(url, self.interface, r.status_code,
-                                                                             r.reason))
-        if len(r.json()) == 0:
-            raise ProvisionerError(
-                'Error no machine with id "{}"'.format(self.machine_id))
+        path = "/api/v1/machine/{}/interface".format(self.machine_id)
+        try:
+            result = self.requester.get(path)
+        except Exception as err:
```
Comment (rengolin) [inline library/get_ip.py]: useless catch & raise (same for all below)

File: library/get_ip.py [hunk 7]
```diff
@@ -4,63 +4,63 @@
...
+            raise err
 
-        return r.json()
+        return result
 
     def get_ip(self):
```
Comment (rengolin) [inline library/get_ip.py]: Can you common up the interface check loop into a local method? and then use it on all other methods.

File: library/common.py (hunk 1 above)
Comment (BaptisteGerondeau) [reply library/common.py]: It sounds charismatic ! Bah, changed it to URLhandler

File: library/common.py (hunk 4 above)
Comment (BaptisteGerondeau) [reply library/common.py]: Now the Requester 'pings' MrP at initialization.

File: library/get_ip.py (hunk 7 above)
Comment (BaptisteGerondeau) [reply library/get_ip.py]: This is now the URLhandler::get_field_by_interface method.

File: library/common.py (hunk 1 above)
Comment (rengolin) [reply library/common.py]: Please, also change all "requester" variables and arguments to "handler".

File: library/common.py [hunk 8]
```diff
@@ -5,25 +5,121 @@
...
+
+
+class URLhandler(object):
+    def __init__(self, mrp_url, mrp_token):
+        try:
+            self.base_url = mrp_url
+            self.headers = {'Authorization': mrp_token}
+            self.get("/api/v1/machine?show_all=false")
+        except Exception as err:
+            raise ClientError("Invalid URL or token for MrP")
```
Comment (rengolin) [inline library/common.py]: Can you include the 'err' in the message, too? To help understand what the error actually is (proxy? no connection? bad url? invalid token? empty strings?)

File: library/get_ip.py [hunk 9]
```diff
@@ -4,63 +4,46 @@
...
-        url = urljoin(self.mrp_url,
-                      "/api/v1/machine/{}/interface".format(self.machine_id))
-        r = requests.get(url, headers=headers)
-        if r.status_code != 200:
-            raise ProvisionerError('Error fetching {} {}, HTTP {} {}'.format(url, self.interface, r.status_code,
-                                                                             r.reason))
-        if len(r.json()) == 0:
-            raise ProvisionerError(
-                'Error no machine with id "{}"'.format(self.machine_id))
+    def find_field_by_interface(self, listOfDicts, identifier, match, return_field):
```
Comment (rengolin) [inline library/get_ip.py]: Hum, a bit too far this time. :)

listOfDicts == self.get_interface() and match = self.interface, both class variables that don't need to be passed.

identifier == 'identifier' for all cases, doesn't need to be passed either.

Right now, we only have interfaces, so no need to be this generic. If/when we have more, we can make it more generic, but not now.

File: library/common.py (hunk 8 above)
Comment (BaptisteGerondeau) [reply library/common.py]: In Python all exceptions include the previous ones (in the stack) in the '__context__' attribute.
But to make it more explicit I'll use "raise a from b" which makes a.__cause__ = b, thus leading to a more explicit output.

Comment (rengolin) [reply library/common.py]: Good point.

**Comment by BaptisteGerondeau** [comment]
@rengolin

**Comment by BaptisteGerondeau** [comment]
@rengolin

**Pull request merged**